Fixed Issue #6966 [Status bar overlap on iPads using Tailwind CSS]

Xeno-Tek opened the pull request:
# Fix: Status Bar Overlap on iPads

## Overview

This pull request resolves an issue where content overlaps with the status bar on iPad devices. The fix utilizes Tailwind CSS classes to account for the status bar height and safe area insets on iPads without affecting other devices.

## Related Issue

Closes #6966

## Changes

### `tailwind.config.cjs`

- **Added** iPad-specific breakpoints.
- **Introduced** a new spacing value `ipad-status-bar` for the iPad status bar height.

### `App.tsx`

- **Updated** the main container to include `ipad:pt-ipad-status-bar`, adding proper top padding on iPads.

### `Header.tsx`

- **Adjusted** the header component to account for the status bar on iPads using iPad-specific Tailwind classes.

## Screenshots

### iPad 10th Generation (Portrait)

<img src="https://github.com/user-attachments/assets/4f8f9b95-485a-4060-a692-9fdd7cfe7949" width="300">

### iPad 10th Generation (Landscape)

<img src="https://github.com/user-attachments/assets/8d2bfa57-a449-44a9-a9b8-08e7c107d9d2" width="300">

### iPhone 16 Pro Max

<img src="https://github.com/user-attachments/assets/114ad2b9-3fe8-4b5b-b491-c8e61a25217c" width="300">

### iPhone SE 3rd Generation

<img src="https://github.com/user-attachments/assets/d12deb69-0647-4b3d-acba-ccd2f38c23fe" width="300">

### iPad Pro 13-inch (M4)

<img src="https://github.com/user-attachments/assets/f6c02d9b-0575-414e-af7b-53019433a004" width="300">

## Testing

- [x] Verified that the status bar no longer overlaps with content on iPad devices.
- [x] Tested in both portrait and landscape orientations on iPads.
- [x] Ensured the header and main content are properly positioned below the status bar on iPads.
- [x] Used only Tailwind CSS classes for styling adjustments.

Xeno-Tek commented:
@VIKTORVAV99, 
I've implemented the changes using Tailwind CSS to address the status bar overlap issue. Please let me know If this works.

VIKTORVAV99 commented:
@Xeno-Tek are those new screenshots?
They still show that the map is not behind the status bar / safe area. Which we wanted to keep.

Xeno-Tek commented:
@VIKTORVAV99 Thanks for pointing that out! I initially thought it was acceptable for the map not to overlay the safe area insets, but I've gone ahead and adjusted it as per your feedback. The map now correctly overlays the safe area insets. 


1. Updated `main.tsx`:
   * Added CSS variables for safe area insets.
   * Created new utility classes for extending content to the top edge and handling content within safe areas.
   * Added padding to the root element to account for the safe area top inset.
2. Modified `MapWrapper.tsx`:
   * Added the `extend-to-top-edge` class to the map container div to allow it to extend behind the status bar.
3. Updated `ZoomControls.tsx`:
   * Implemented dynamic top margin calculation based on the safe area top inset.
   * Used a `useEffect` hook to update the top margin on component mount and window resize.
   
## Preview Images

### iPad 10th Generation (Portrait)
<img src="https://github.com/user-attachments/assets/22b338ec-595c-4987-8312-69e267005811" width="300" alt="iPad 10th Generate Fix safe area overlap">

### iPad 10th Generation (Landscape)
<img src="https://github.com/user-attachments/assets/09dc5490-b6cd-4d99-ba47-f243dc4dffbf" width="400" alt="iPad 10th gen fix horizontaly">

### iPhone 16 Pro Max
<img src="https://github.com/user-attachments/assets/35713a4f-4999-4dee-88c3-3be12b024f85" width="250" alt="Iphone 16  Pro Max fix">

### iPhone SE 3rd Generation
<img src="https://github.com/user-attachments/assets/c57955e3-2089-4e18-bf5e-8774988c29aa" width="200" alt="Iphone SE 3rd gen fix">

### iPad Pro 13-inch M4
<img src="https://github.com/user-attachments/assets/b136d067-aa89-4365-9de0-4e160f6b5e80" width="400" alt="iPad Pro 13-inch M4">

VIKTORVAV99 commented on mobileapp/ios/App/Podfile:
```diff
@@ -12,7 +12,6 @@ def capacitor_pods
   pod 'Capacitor', :path => '../../node_modules/.pnpm/@capacitor+ios@6.0.0_@capacitor+core@6.0.0/node_modules/@capacitor/ios'
   pod 'CapacitorCordova', :path => '../../node_modules/.pnpm/@capacitor+ios@6.0.0_@capacitor+core@6.0.0/node_modules/@capacitor/ios'
   pod 'CapacitorApp', :path => '../../node_modules/.pnpm/@capacitor+app@6.0.0_@capacitor+core@6.0.0/node_modules/@capacitor/app'
-  pod 'CapacitorShare', :path => '../../node_modules/.pnpm/@capacitor+share@6.0.2_@capacitor+core@6.0.0/node_modules/@capacitor/share'
```
You can't simply remove these, the app will break if you do...

Xeno-Tek replied:
My bad for the mix-up, thought they would not get pushed due to git ignore.

VIKTORVAV99 commented on web/src/main.tsx:
```diff
@@ -18,6 +18,45 @@ import enableErrorsInOverlay from 'utils/errorOverlay';
 
+const style = document.createElement('style');
+style.textContent = `
+  :root {
+    --sat: env(safe-area-inset-top);
+    --sar: env(safe-area-inset-right);
+    --sab: env(safe-area-inset-bottom);
+    --sal: env(safe-area-inset-left);
+  }
```
are you not simply re-defining existing envs here?

Why do we need this?

VIKTORVAV99 commented on web/src/main.tsx:
```diff
@@ -18,6 +18,45 @@ import enableErrorsInOverlay from 'utils/errorOverlay';
+    --sal: env(safe-area-inset-left);
+  }
+  
+  body {
+    padding: 0;
+    margin: 0;
+  }
+  
+  #root {
+    padding-top: var(--sat);
+  }
+  
+  .extend-to-top-edge {
+    position: absolute;
+    top: 0;
+    left: 0;
+    right: 0;
+    height: 100%;
+    margin-top: calc(var(--sat) * -1);
+  }
+  
+  .content-within-safe-area {
+    position: relative;
+    z-index: 1;
+    padding: var(--sar) var(--sab) var(--sal);
+  }
+  
+  .pad-for-header {
+    padding-top: var(--sat);
+  }
+`;
```
We already have a index.css file for styles like this. If they are in typescript files they should be tailwind classes, otherwise they should go into the CSS file as these are universal.

VIKTORVAV99 commented on mockserver/package-lock.json:
This file should not exist, we use PNPM.

VIKTORVAV99 commented on mobileapp/ios/App/Podfile.lock:
```diff
@@ -13,6 +13,7 @@ DEPENDENCIES:
   - "CapacitorCordova (from `../../node_modules/.pnpm/@capacitor+ios@6.0.0_@capacitor+core@6.0.0/node_modules/@capacitor/ios`)"
   - "CapacitorShare (from `../../node_modules/.pnpm/@capacitor+share@6.0.2_@capacitor+core@6.0.0/node_modules/@capacitor/share`)"
 
+
```
Avoid diffs in unrelated files please.

VIKTORVAV99 commented on web/src/features/map-controls/ZoomControls.tsx:
```diff
@@ -1,15 +1,30 @@
-  const isConsumptionOnlyMode = useFeatureFlag('consumption-only');
-  const marginTop = isConsumptionOnlyMode ? 54 : 98;
+  const [topMargin, setTopMargin] = useState(98);
+  useEffect(() => {
+    const updateTopMargin = () => {
+      const safeAreaTop =
+        Number.parseInt(
+          getComputedStyle(document.documentElement).getPropertyValue('--sat'),
+          10
+        ) || 0;
+      setTopMargin(98 + safeAreaTop);
+    };
+
+    updateTopMargin();
+    window.addEventListener('resize', updateTopMargin);
+    return () => window.removeEventListener('resize', updateTopMargin);
+  }, []);
```
This should use a the resizeObserverHook, useEffects should be options of last resort in react. Using a event listener instead of a resizeObserver also have a performance overhead.

VIKTORVAV99 commented on web/src/features/map-controls/ZoomControls.tsx:
```diff
@@ -1,15 +1,30 @@
       style={{
         marginRight: 12,
-        marginTop: marginTop,
+        marginTop: topMargin,
```
Why this change?

Please try and keep the diffs to a minimum.

VIKTORVAV99 commented on web/src/features/map/MapWrapper.tsx:
```diff
@@ -20,7 +20,13 @@ export default function MapWrapper() {
+      {shouldShowFallback ? (
+        <MapFallback />
+      ) : (
+        <div className="extend-to-top-edge absolute inset-0">
+          <Map />
+        </div>
```
Can this not be handled inside the <Map> component?
We want to avoid creating more elements than is needed.

VIKTORVAV99 reviewed (changes requested):
While this seems to work it does this by adding duplicate styles which will be hard to debug later.

There is also files changed in this PR that should not be changed at all which needs to be cleaned up before merging.

Xeno-Tek commented:
@VIKTORVAV99 Ive made the changes according to your feedback, please check the files changed.


## Test Results After Implementation of Changes

### iPad 10th Gen - Portrait Mode

<img src="https://github.com/user-attachments/assets/1a848f8e-d51d-4c8a-a614-79a3f1a7fe78" alt="iPad 10th Gen Portrait Mode" width="300"/>

### iPad 10th Gen - Landscape Mode

<img src="https://github.com/user-attachments/assets/56ab9476-a4be-4463-ab0d-c960038f244a" alt="iPad 10th Gen Landscape Mode" width="300"/>

### iPhone 16 Pro Max

<img src="https://github.com/user-attachments/assets/47797b7e-c593-4d35-8cac-30c06002ae55" alt="iPhone 16 Pro Max" width="300"/>

VIKTORVAV99 commented:
Please stop merging master into this. I get a notification every time you do...

Xeno-Tek commented:
@VIKTORVAV99 @tonypls, could you please let me know when the PR will be reviewed? I'm working on this as part of my assignment, so I won't be able to contribute or make fixes after my report is completed.
